crates/fuel-core/src/service/adapters/gas_price_adapters.rs

```diff
@@ -39,6 +57,34 @@ impl L2Data for OnChainIterableKeyValueView {
     }
 }
 
+impl GasPriceData for Database<GasPriceDatabase> {
+    fn get_metadata(
+        &self,
+        block_height: &BlockHeight,
+    ) -> StorageResult<Option<UpdaterMetadata>> {
+        self.storage::<GasPriceMetadata>()
+            .get(block_height)
+            .map(|metadata| metadata.map(|metadata| metadata.as_ref().clone()))
+    }
+
+    fn set_metadata(&mut self, metadata: &UpdaterMetadata) -> StorageResult<()> {
+        let height = metadata.l2_block_height();
+        let mut tx = self.write_transaction();
+        tx.storage_as_mut::<GasPriceMetadata>()
+            .insert(&height, metadata)?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    fn latest_height(&self) -> Option<BlockHeight> {
+        HistoricalView::latest_height(self)
+    }
+
+    fn rollback_last_block(&self) -> StorageResult<()> {
```

Comment:
I think we don't need to leak this kind of functionality to the service.

I think it would be better if `FuelService` were responsible for database integrity. It is okay if other databases are behind than on-chain database. It is the responsibility of the service to catch up(since only the service knows how to advance the state). But if the on-chain database is behind other databases, then `FuelService` should roll back them since only `FuelService` knows that the database supports the state rewind feature.

Reply:
Hmmmm. Kinda confused. InitializationTask currently handles that and I thought we wanted to get rid of it and move that code into `GasPriceService`.

I think @xgreenx  right that it's leaking something, but at the same time we have to leak the idea that they can be out of sync to include the syncing logic into `GasPriceService`.

Reply:
i.e. I'm not sure how the `FuelService` could handle it without accessing the metadata and updater code.

Reply:
I think @xgreenx here means that the `FuelService` *should* be aware of rollback strategy for each of the db's included within `CombinedDatabase`, which isn't that complex at the moment where we're leaking business logic, its just rolling back to the same height as the `OnChainDb`

Reply:
addressed in 9575a21b48
seems cleaner than before :)

Reply:
slightly unsure about where you'd like to place the rewind/rollback call ~ i've placed it before the `init_sub_services` call :)